src/compiler/lowering/SpecializeGenericFunctionsPass.cpp

```diff
@@ -188,6 +188,7 @@ namespace {
         std::multimap<std::string, func::FuncOp> specializedVersions;
         std::set<func::FuncOp> visited;
         std::set<func::FuncOp> called;
+        std::set<func::FuncOp> templateFunctions;
```

Comment:
Initially, I thought the check `isFunctionTemplate` would already do the trick, but apparantely it does something else. I basically remove the "template function", which is specialized from considerations in later passes (i.e., doing a type inference on such a function doesn't make much sense imho). Template functions are those, which are specialized, that's why I add them unconditionally. According to the test suite this should be fine, but I feel this will require more thorough testing (independent from the reported bug).

In general there are quite a few question marks in this pass. For example the constant parameter propagation leads to X functions with identical code (except for the constant parameter of course) when invoking the UDF X times.

Reply:
I've played around with the code a bit now and see what you mean. The generation of a separate function for each invocation (even if it is the same parameter) is a bit weird and could be improved.